### PR TITLE
[ROCm] [ROCm2.5+] switch to rocThrust for thrust/cub APIs

### PR DESCRIPTION
**Pull request by iotamudelta**
Enable rocThrust with hipCUB and rocPRIM for ROCm. They are the ROCm implementations of the thrust and cub APIs and replace the older hip-thrust and cub-hip packages going forward. ROCm 2.5 is the first release to contain the new packages as an option, as of 2.6 they will be the only available option.

Add hipification rules to correctly hipify thrust::cuda to thrust::hip and cub:: to hipcub:: going forward. Add hipification rules to hipify specific cub headers to the general hipcub header.

Infrastructure work to correctly find, include and link against the new packages. Add the macro definition to choose the HIP backend to Thrust.

Since include chains are now a little different from CUDA's Thrust, add includes for functionality used where applicable.

Skip four tests that fail with the new rocThrust for now.

### CODE REVIEW

**Comment by iotamudelta** [comment]
@bddppq we will need to test this on newer docker images.

File: cmake/public/LoadHIP.cmake [hunk 1]
```diff
@@ -150,6 +150,10 @@ IF(HIP_FOUND)
   #find_package_and_print_version(hipsparse REQUIRED)
   find_package_and_print_version(rocsparse REQUIRED)
 
+  find_package(rocprim REQUIRED CONFIG PATHS "/opt/rocm/rocprim")
```
Comment (bddppq) [inline cmake/public/LoadHIP.cmake]: Better to follow the convention of other rocm libraries (i.e. define `ROCPRIM_PATH`, `rocprim_DIR` and call `find_package_and_print_version`

File: cmake/Dependencies.cmake [hunk 2]
```diff
@@ -871,6 +871,7 @@ if(USE_ROCM)
     list(APPEND HIP_CXX_FLAGS -Wno-unused-command-line-argument)
     list(APPEND HIP_CXX_FLAGS -Wno-duplicate-decl-specifier)
     list(APPEND HIP_CXX_FLAGS -DCAFFE2_USE_MIOPEN)
+    list(APPEND HIP_CXX_FLAGS -DTHRUST_DEVICE_SYSTEM=THRUST_DEVICE_SYSTEM_HIP)
```
Comment (bddppq) [inline cmake/Dependencies.cmake]: what does this flag do?

File: aten/src/THC/generic/THCTensorMode.cu [hunk 3]
```diff
@@ -1,3 +1,6 @@
+#if defined(__HIP_PLATFORM_HCC__)
+#include <thrust/iterator/constant_iterator.h>
```
Comment (bddppq) [inline aten/src/THC/generic/THCTensorMode.cu]: Why is here conditionally included but not in `aten/src/ATen/native/cuda/EmbeddingBag.cu`?

**Comment by bddppq** [comment]
Building new docker images here https://ci.pytorch.org/jenkins/job/caffe2-docker-trigger/289/

File: cmake/Dependencies.cmake (hunk 2 above)
Comment (iotamudelta) [reply cmake/Dependencies.cmake]: See PR message: Add the macro definition to choose the HIP backend to Thrust.

**Comment by bddppq** [comment]
cc @xw285cornell this will require update in tp2 as well

File: aten/src/THC/generic/THCTensorMode.cu (hunk 3 above)
Comment (iotamudelta) [reply aten/src/THC/generic/THCTensorMode.cu]: This one had no thrust includes - EmbeddingBag already had a bunch so it looked like an oversight. More than happy to remove the conditionals here. Shall I?

**Comment by bddppq** [comment]
Testing with new docker images:
https://ci.pytorch.org/jenkins/job/caffe2-builds/job/py2-clang7-rocmdeb-ubuntu16.04-trigger-test/31998/
https://ci.pytorch.org/jenkins/job/caffe2-builds/job/py2-devtoolset7-rocmrpm-centos7.5-trigger-test/28703/
https://ci.pytorch.org/jenkins/job/pytorch-builds/job/py2-clang7-rocmdeb-ubuntu16.04-trigger/35656/
https://ci.pytorch.org/jenkins/job/pytorch-builds/job/py2-devtoolset7-rocmrpm-centos7.5-trigger/17244/

**Comment by iotamudelta** [comment]
Looks like PyTorch tests passed and we need to disable `TestUniqueOps` for the Caffe2 tests for now.

**Review by ezyang — DISMISSED** [review]
Looks reasonable enough. Not sure about the failures though

**Review by bddppq — COMMENTED** [review]
I have discussed with @xw285cornell offline, we prefer to delay this switch to future official release version:
1. Per our last meeting with amd, we were recommended to not use rocThrust in ROCm2.5
2. There are non-trivial amount of work to update ROCm toolchain in our internal infra. Considering there are no major fixes in the thrust library in 2.5, we would prefer to batch the update work to next release version that comes with more major fixes.

**Comment by bddppq** [comment]
cc @gottbrath

**Comment by bddppq** [comment]
The caffe2 unique_ops issue is caused by [`THRUST_VERSION`](https://github.com/pytorch/pytorch/blob/db5a3c2/caffe2/operators/unique_ops.cu#L28) and that caused the operator being not registered:

```
RuntimeError: [enforce fail at operator.cc:264] op. Cannot create operator of type 'Unique' on the device 'HIP'. Verify that implementation for the corresponding device exist. It might also happen if the binary is not linked with the operator implementation code.
```

**Review by xw285cornell — COMMENTED** [review]
Right, if possible we'll want to delay it. Land it as is will break our internal repository (which we only have hip-thrust). And update it to rocthrust will be some non-trivial amount of work (I can do, but I'm doing some other debugging right now. Unless this one is urgent..)

**Comment by iotamudelta** [comment]
I appreciate that. Our hope was to drop the old Thrust completely with ROCm 2.6 and now (i.e., a few weeks before release) switch PyTorch as a critical consumer over to the new solution already.

**Comment by iotamudelta** [comment]
@bddppq concerning `THRUST_VERSION`, we can include `include/thrust/version.h` explicitly which should define that macro to `100902`.

**Comment by bddppq** [comment]
New triggered runs:
https://ci.pytorch.org/jenkins/job/caffe2-builds/job/py2-clang7-rocmdeb-ubuntu16.04-trigger-test/31999/
https://ci.pytorch.org/jenkins/job/caffe2-builds/job/py2-devtoolset7-rocmrpm-centos7.5-trigger-test/28863/
https://ci.pytorch.org/jenkins/job/pytorch-builds/job/py2-clang7-rocmdeb-ubuntu16.04-trigger/35815/
https://ci.pytorch.org/jenkins/job/pytorch-builds/job/py2-devtoolset7-rocmrpm-centos7.5-trigger/17245/

**Comment by iotamudelta** [comment]
@bddppq do you understand the bench failures for Caffe2?

**Comment by bddppq** [comment]
@iotamudelta yep :-) we have just added two workers to the pool and turns out these boxes have mi50 (instead of mi60), so their gpu memory capacity is not enough to run the corresponding batch size hardcoded in the benchmark.
I'm trying to rerun the benchmark on the right nodes, but unfortunately our jenkins CI is down right now :(. I will re-trigger them once our jenkins is back.

**Comment by iotamudelta** [comment]
@bddppq thanks so much!

**Comment by bddppq** [comment]
Re-triggered bench jobs:
https://ci.pytorch.org/jenkins/job/caffe2-builds/job/py2-clang7-rocmdeb-ubuntu16.04-trigger-test/32000/
https://ci.pytorch.org/jenkins/job/caffe2-builds/job/py2-devtoolset7-rocmrpm-centos7.5-trigger-test/28907/

**Comment by iotamudelta** [comment]
@bddppq looks like everything passed properly.

**Comment by bddppq** [comment]
Yay! I think we can keep this PR open and use it next time we update the rocm toolchain.

**Comment by iotamudelta** [comment]
So you want to keep the CI on ROCm 2.4 for now and directly jump to 2.6?

**Comment by bddppq** [comment]
CI is now on 2.5. Yep we prefer to not do an update until next release with major fixes/improvements

**Review by soumith — CHANGES_REQUESTED** [review]
removing from ready to review queue.
@bddppq feel free to dismiss my review whenever it's time.

**Comment by iotamudelta** [comment]
@bddppq ROCm 2.6 is live now - so we could now get this PR landed together with the upgrade as planned.

**Comment by iotamudelta** [comment]
@bddppq ROCm 2.7 is live now

**Comment by xw285cornell** [comment]
I guess the first thing we need to do is to upgrade the CI machines with the latest driver.

**Comment by iotamudelta** [comment]
@xw285cornell ROCm 2.7 docker will run on ROCm 2.6 kernel driver that is on the CI hosts.

**Comment by xw285cornell** [comment]
I think the common practice is to use the new driver + old runtime (and I believe this is the agreement between us). I'm not sure the other way is very well tested (or does your QA supports that?)

We'll also need a driver upgrade on our side.

**Comment by iotamudelta** [comment]
We had some internal discussion and updating the docker is supported in this case. However, the CI host nodes will also be upgraded.

**Comment by xw285cornell** [comment]
@iotamudelta I've tested this internally and it looks good to land. However I have to add the internal upgrade (it'll be one step instead of two steps in OSS,  this PR + docker image ) with this diff, but I'm not supposed to change this PR. Do you mind if I create another PR?

**Comment by iotamudelta** [comment]
@xw285cornell go right ahead - as long as the CI gets updated and we switch to rocThrust, I am a happy camper and can deal with the merge fallout! :-)

**Pull request closed without merging**